horovod/mxnet/__init__.py

```diff
@@ -47,10 +47,16 @@ def __init__(self, optimizer, gradient_predivide_factor=1.0, num_groups=0, proce
             raise ValueError('gradient_predivide_factor not supported yet with ROCm')
 
         self._optimizer = optimizer
-        # Normalizing rescale_grad by Horovod size, which is equivalent to
-        # performing average in allreduce, has better performance.
-        self._optimizer.rescale_grad *= (gradient_predivide_factor / process_set.size())
         self._gradient_predivide_factor = gradient_predivide_factor
+        self._average_in_framework=False
+        # C++ backend will apply additional 1 / size() factor to postscale_factor for op == Average.
+        self._postscale_factor = self._gradient_predivide_factor
+        if rocm_built() or nccl_built() < 21000:
```

Comment:
It's a bit surprising to see this case distinction on this level. I would have expected it in C++ `horovod_mxnet_allreduce_async()`, but if that's not possible easily, it will be fine like this.

Overall the code path with averaging to be quite convoluted with MXNet. Why do we have more branches in framework-specific code with MXNet than we do with TensorFlow for instance? Does it make Horovod with MXNet faster than with the other frameworks?

Reply:
The reason for this is because MXNet, unlike the other frameworks, has the ability to rescale gradients in the optimizer (via `rescale_grad`), usually with the scaling fused into the weight update kernel. In the past, this was a nice performance optimization because Horovod required calling a separate kernel for scaling gradients (or even worse, called a kernel per individual gradient for scaling). By using MXNet's gradient rescaling functionality, we were able to cut down on the number of kernels launched and improve performance. Since I fused scaling into the buffer packing, this has been of less benefit; however, I maintained the older path for platforms (e.g. ROCM) that didn't have the CUDA kernels I wrote to do the scaling/packing and could still benefit from the gradient scaling in the optimizer. Perhaps this is no longer needed since #3486, but it is hard to say without any CI testing of that platform.

Reply:
Ah, thank you, I didn't know the `rescale_grad` feature would be that powerful. That's useful!